cleanrl/diayn_sac_continuous_action.py

```diff
@@ -0,0 +1,636 @@
+                        target_param.data.copy_(self.args.tau * param.data + (1 - self.args.tau) * target_param.data)
+                    for param, target_param in zip(self.qf2.parameters(), self.qf2_target.parameters()):
+                        target_param.data.copy_(self.args.tau * param.data + (1 - self.args.tau) * target_param.data)
+
+                if global_step % 100 == 0:
+                    self.writer.add_scalar("losses/qf1_values", qf1_a_values.mean().item(), global_step)
+                    self.writer.add_scalar("losses/qf2_values", qf2_a_values.mean().item(), global_step)
+                    self.writer.add_scalar("losses/qf1_loss", qf1_loss.item(), global_step)
+                    self.writer.add_scalar("losses/qf2_loss", qf2_loss.item(), global_step)
+                    self.writer.add_scalar("losses/qf_loss", qf_loss.item() / 2.0, global_step)
+                    self.writer.add_scalar("losses/actor_loss", actor_loss.item(), global_step)
+                    self.writer.add_scalar("losses/alpha", self.alpha, global_step)
+                    self.writer.add_scalar("losses/discriminator_loss", discriminator_loss.item(), global_step)
+                    print("SPS:", int(global_step / (time.time() - start_time)))
+                    self.writer.add_scalar("charts/SPS", int(global_step / (time.time() - start_time)), global_step)
+                    if self.args.autotune:
+                        self.writer.add_scalar("losses/alpha_loss", alpha_loss.item(), global_step)
+                if global_step % 10000 == 0:
+                    self.save_model()
+        return
+
+    def use_skills(self):
+        """
+        Use pre-trained skills to train the agent.
+        """
+        low = np.hstack([self.envs.single_observation_space.low, np.full(args.num_skills, 0)])
+        high = np.hstack([self.envs.single_observation_space.high, np.full(args.num_skills, 1)])
+        aug_obs_space = gym.spaces.Box(low=low, high=high)
+        rb = ReplayBuffer(
+            args.buffer_size,
+            aug_obs_space,
+            self.envs.single_action_space,
+            device,
+            handle_timeout_termination=True,
+        )
+
+        start_time = time.time()
+
+        # Evaluate each skill to find the best performing skill
+        if self.args.best_skill is not None:
+            best_skill = self.args.best_skill
+        else:
+            print("Best skill not provided. Evaluating each skill to find the best performing skill.")
+            scores = self.evaluate_skills(eval_episodes=1)
+            best_skill = np.argmax(scores)
+
+        # initializing the networks with best skill and the fine-tuning on it
+        one_hot_z = np.zeros((1, self.args.num_skills), dtype=np.float32)
+        one_hot_z[0, best_skill] = 1
+
+        # TRY NOT TO MODIFY: start the game
+        obs = self.envs.reset()
+        z_aug_obs = aug_obs_z(obs, one_hot_z)
+        for global_step in range(self.args.total_timesteps):
```

Comment:
What is the difference between `learn_skills` and `use_skills`? Why are both of them going over `1000000` steps?

Reply:
`learn_skills` is the unsupervised skill learning phase, whereas `use_skills` is to fine-tune the trained model to optimize for the environment reward